File: src/main/java/x/withlithum/neoware/data/content/packs/ContentPackLoader.java
/*
 * SPDX-FileCopyrightText: 2025-2026 WithLithum & contributors
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

package x.withlithum.neoware.data.content.packs;

import com.google.gson.JsonParser;
import com.google.gson.JsonSyntaxException;
import lombok.extern.slf4j.Slf4j;
import net.minestom.server.codec.Result;
import net.minestom.server.codec.Transcoder;
import org.jspecify.annotations.NullMarked;
import org.jspecify.annotations.Nullable;
import x.withlithum.neoware.data.content.hierarchy.ContentTree;
import x.withlithum.neoware.data.content.hierarchy.ContentTreeFactory;
import x.withlithum.neoware.util.results.NeoResult;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

@NullMarked
@Slf4j
public final class ContentPackLoader {
    private ContentPackLoader() {
        throw new AssertionError("No ContentPackLoader instances for you!");
    }

    public static final String PACK_META_FILE = "neoware_pack.json";
    public static final String PACK_DATA_DIR = "data";

    public static <V extends ContentTree<V>> V loadAllMerged(Path path,
                                                             ContentTreeFactory<V> factory) {
        if (!Files.isDirectory(path)) {
            log.warn("'{}' is not a directory", path);
            return factory.getEmpty();
        }

        var loadCount = 0;
        V tree = factory.getEmpty();
        try {
            try (final var s = Files.list(path)
                .filter(Files::isDirectory)) {
                final var list = s.toList();

                for (var f : list) {
                    if (!Files.isDirectory(f)) {
                        continue;
                    }

                    final var pack = loadOne(f, factory);
                    switch (pack) {
                        case NeoResult.Error<V> e -> e.logWarn(log);
                        case NeoResult.Ok<V> o -> {
                            tree = tree.merge(o.getValue());
                            loadCount++;
                        }
                        default -> throw new IllegalStateException("Impossible state");
                    }
                }
            }
        } catch (JsonSyntaxException | IOException e) {
            log.warn("Failed to load content tree", e);
        }

        log.info("Loaded {} content packs", loadCount);
        return tree;
    }

    public static <V extends ContentTree<V>> NeoResult<V> loadOne(Path path,
                                                                  ContentTreeFactory<V> factory) {
        if (!Files.isDirectory(path)) {
            throw new IllegalArgumentException("The provided path is not a directory.");
        }

        final var packMeta = loadPackMeta(path);
        if (packMeta == null) {
            return new NeoResult.Error<>("Failed to load pack meta", null);
        }
        if (packMeta.dataVersion() != factory.getDataVersion()) {
            return new NeoResult.Error<>("Unsupported pack version", null);
        }

        final var dataPath = path.resolve(PACK_DATA_DIR);
        if (!Files.isDirectory(dataPath)) {
            return new NeoResult.Error<>("Data path is not a directory or does not exist", null);
        }

        return factory.loadDir(path);
    }

    private static @Nullable ContentPackMeta loadPackMeta(Path packBase) {
        final var metaFile = packBase.resolve(PACK_META_FILE);
        if (!Files.isRegularFile(metaFile)) {
            log.warn("Cannot find meta file '{}', metaFile", metaFile);
            return null;
        }

        try (final var reader = Files.newBufferedReader(metaFile)) {
            final var result = ContentPackMeta.CODEC.decode(Transcoder.JSON,
                JsonParser.parseReader(reader));
            switch (result) {
                case Result.Ok<ContentPackMeta> ok -> { return ok.value(); }
                case Result.Error<ContentPackMeta> err -> {
                    log.warn("Failed to decode meta file '{}': {}", metaFile, err.message());
                    return null;
                }
            }

        } catch (JsonSyntaxException | IOException e) {
            log.warn("Failed to read meta file", e);
            return null;
        }
    }
}

File: src/main/kotlin/x/withlithum/neoware/data/content/packs/ContentPackLoader.kt
/*
 * SPDX-FileCopyrightText: 2025-2026 WithLithum & contributors
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

package x.withlithum.neoware.data.content.packs

import io.github.oshai.kotlinlogging.KotlinLogging
import io.github.wasabithumb.jtoml.JToml
import okio.FileSystem
import okio.Path
import x.withlithum.neoware.data.content.hierarchy.ContentTree
import x.withlithum.neoware.data.content.hierarchy.ContentTreeFactory
import x.withlithum.neoware.data.encode.TomlTranscoder
import x.withlithum.neoware.util.io.isDirectory
import x.withlithum.neoware.util.io.isRegularFile
import x.withlithum.neoware.util.io.read
import net.minestom.server.codec.Result as MResult

object ContentPackLoader {
    const val PACK_META_FILE = "neoware_pack.toml"
    const val PACK_DATA_DIR = "data"

    private val logger = KotlinLogging.logger {}
    private val toml = JToml.jToml()

    fun <V> loadAll(path: Path, fs: FileSystem, factory: ContentTreeFactory<V>): V where V : ContentTree<V> {
        if (!fs.isDirectory(path)) {
            return factory.empty
        }

        var loadCount = 0
        var tree: V? = null
        fs.list(path).forEach f@{
            if (!fs.isDirectory(it)) {
                return@f
            }

            val pack = loadPack(it, fs, factory) ?: return@f
            tree = tree?.merge(pack.tree) ?: pack.tree

            loadCount++
        }

        logger.info { "Loaded $loadCount content packs" }
        return tree ?: factory.empty
    }

    fun <V> loadPack(
        path: Path,
        fs: FileSystem,
        factory: ContentTreeFactory<V>
    ): ContentPack<V>? where V : ContentTree<V> {
        if (!fs.isDirectory(path)) {
            throw IllegalArgumentException("'$path' is not a directory.")
        }

        val packMeta = loadPackMeta(path, fs) ?: return null
        if (packMeta.dataVersion != factory.dataVersion) {
            logger.warn { "Skipping '${path.name}'; pack data version is ${packMeta.dataVersion} but only ${factory.dataVersion} is supported" }
            return null
        }

        val dataPath = path.resolve(PACK_DATA_DIR)
        val tree: V
        if (!fs.isDirectory(dataPath)) {
            logger.warn { "Empty pack '${path.name}': data path is not a directory" }
            tree = factory.empty
        } else {
            tree = factory.loadDir(dataPath, fs)
        }

        return ContentPack(packMeta, tree)
    }

    private fun loadPackMeta(packPath: Path, fs: FileSystem): ContentPackMeta? {
        val packMeta = packPath.resolve(PACK_META_FILE)
        if (!fs.isRegularFile(packMeta)) {
            logger.warn { "Pack directory $packMeta has no metadata file" }
            return null
        }

        val tomlFile = toml.read(packMeta, fs)
        when (val tomlResult = ContentPackMeta.CODEC.decode(TomlTranscoder.INSTANCE, tomlFile)) {
            is MResult.Error -> {
                logger.warn { "Failed to decode pack metadata: ${tomlResult.message}" }
                return null
            }

            is MResult.Ok -> return tomlResult.value
        }
    }
}